src/charm.py

```diff
@@ -122,51 +178,56 @@ def mysql(self) -> MySQL:
         """Returns MySQL object"""
         mysql_config = {
             "app_name": self.model.app.name,
-            "host": self.hostname,
+            "host": self.unit_ip,
             "port": self.model.config["port"],
             "user_name": "root",
-            "mysql_root_password": self._stored.mysql_setup[
-                "MYSQL_ROOT_PASSWORD"
-            ],
+            "mysql_root_password": self.mysql_root_password,
         }
         return MySQL(mysql_config)
 
     @property
-    def hostname(self) -> str:
-        """Unit hostname"""
-        unit_id = self.unit.name.split("/")[1]
-        return "{0}-{1}.{0}-endpoints".format(self.model.app.name, unit_id)
+    def unit_ip(self) -> str:
+        """Returns unit's IP"""
+        return str(self.model.get_binding(PEER).network.bind_address)
 
     @property
     def mysql_root_password(self) -> Union[str, None]:
         """
-        This property return MYSQL_ROOT_PASSWORD from StoredState.
-        If the password isn't in StoredState, generates one.
+        This property returns MYSQL_ROOT_PASSWORD from the config,
+        if the password isn't in StoredState, generates one.
         """
 
-        if not self.unit.is_leader():
-            return None
-
-        if "MYSQL_ROOT_PASSWORD" not in self._stored.mysql_setup:
-            self._stored.mysql_setup[
+        config = self.model.config
+        if config.get("MYSQL_ROOT_PASSWORD"):
+            logger.debug("Getting the root password from config")
+            self._stored.mysql_setup["MYSQL_ROOT_PASSWORD"] = config[
                 "MYSQL_ROOT_PASSWORD"
-            ] = MySQL.new_password(20)
+            ]
+            return self._stored.mysql_setup["MYSQL_ROOT_PASSWORD"]
+
+        if self.unit.is_leader():
+            if not self._stored.mysql_setup["MYSQL_ROOT_PASSWORD"]:
+                self._stored.mysql_setup[
+                    "MYSQL_ROOT_PASSWORD"
+                ] = MySQL.new_password(20)
+                logger.debug("Password generated.")
```

Comment:
This may want to be elevated from debug? Seems like something you'd always want to know

Reply:
Sounds reasonable.
Done